### DISCUSSION TITLE
Accessing files which are part of "install-time"-asset-packs on Android

### DISCUSSION THREAD

**Discussion by dbartsch-bf** [airsdk/Adobe-Runtime-Support#1036] [Q&A]
Hi everyone!
Is there a specific way to access files which are part of "install-time"-assets-packs on Android?
Using URLLoader didn't seem to work during my tests because of io errors. (See Example Code 1)
According to Google's documentation Android's asset manager should be used to access these files.
Are there any AIR classes that use this asset manager internally?

I'm using air_33.1.1.533 and the following entry in the descriptor:
```
<assetPacks>
	<assetPack id="MainAssetPack" delivery="install-time" folder="MainAssetPack" />
</assetPacks>
```

I also checked the aab in 7-Zip and the asset-pack seems to be properly packed.

For the file "MainAssetPack/foobar/test.xml" I am currently using the following path to access the file:
```
foobar/test.xml
```

Example Code 1:
```
var myLoader:URLLoader = new URLLoader();
myLoader.dataFormat = URLLoaderDataFormat.BINARY;
myLoader.addEventListener(Event.COMPLETE, onXMLComplete);
myLoader.addEventListener(IOErrorEvent.IO_ERROR, onIOError);
myLoader.load(new URLRequest(path));
```

**Comment by ajwfrost** [comment]
We've been developing an ANE to access files from asset packs .. I think it's pretty much ready for release, I'll perhaps put the initial version up here whilst we work out how to deploy it via the new AIR Package Manager...

(watch this space - will tag you when I upload it, hopefully today/tomorrow...)

**Reply by dbartsch-bf** [reply]
That would be amazing! :) Thanks a lot for the quick reply ajwfrost! 👍

**Reply by marchbold** [reply]
@ajwfrost If you shoot it my way I'll get it up on the package repository for you.

**Reply by ajwfrost** [reply]
Thanks @marchbold  - it would be good to first get a bit of real-world confirmation that it actually works ... @dbartsch-bf please see below (needs to be renamed from the .zip file i.e. just change .zip to .ane and you can use it...)
[PlayAssetDelivery.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/6931072/PlayAssetDelivery.zip)

If you have any issues/comments on it, please let us know!
thanks

**Reply by dbartsch-bf** [reply]
@ajwfrost Thank you so much for providing the ANE! :)
I will try to integrate the extension today and will let you know if I have any issues.

**Reply by dbartsch-bf** [reply]
@ajwfrost I've encountered some errors while trying to build the AAB during my tests.

1) At first I added the following entry to our descriptor:
`<extensionID>com.harman.air.extension.PlayAssetDeliveryAne</extensionID>`

2) Next I started our build process which resulted in the following errors in the ":app:checkReleaseDuplicateClasses"-task:
[playassetdeliveryane-build-error-01.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/6931650/playassetdeliveryane-build-error-01.txt)

3) As the errors were related to "androidx.appcompat" I removed Distriqt's "androidx.appcompat.ane" from our references
which resulted in the following errors in the ":app:bundleReleaseResources"-task:
[playassetdeliveryane-build-error-02.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/6931663/playassetdeliveryane-build-error-02.txt)

I also tried to update the ANEs from Distriqt's ANE-AndroidSupport to the most recent version but this still resulted in errors with or without referencing "androidx.appcompat.ane". We also use several other Distriqt extensions. Just let me know if you need a complete list or any other additional infos.

**Reply by ajwfrost** [reply]
Hi @dbartsch-bf - thanks for this. So, we have the issue with duplicate classes which I think we may be able to get rid of (not sure whether that was in the latest release that just came out...)

The resource error may be a compatibility issue between different Android versions/libraries. There are others with similar issues e.g.
https://stackoverflow.com/questions/60789109/cannot-solve-error-attribute-alpha-not-found
and a few folk are suggesting that it can be caused by a cache issue (particularly if you've changed the settings but it's the same project...)

Can you perhaps re-try with the latest SDK that's now available at https://airsdk.harman.com/download since we had updated this to ensure gradle doesn't use the cache, to avoid some other issues that had been found...

thanks

**Reply by dbartsch-bf** [reply]
Hi @ajwfrost!
Thanks for the help! I made some progress but I had to disable some other ANEs along the way which we still need.

1) At first I switched to the latest air_33.1.1.575 SDK:
=> Still got "attribute not found"-errors mostly for com.google.android.material.

2) So I also removed the reference to "com.google.android.material.ane":
=> The errors went down to this single error:
```
> Android resource linking failed
  AAPT: error: resource style/Theme.AppCompat.NoActionBar (aka com.traviangames.railnationworld:style/Theme.AppCompat.NoActionBar) not found.
  error: resource style/Theme.AppCompat.Dialog (aka com.traviangames.railnationworld:style/Theme.AppCompat.Dialog) not found.
  error: failed linking references.
```

3) I remembered something about Facebook using Theme.AppCompat.NoActionBar in the back of my head, so I removed the references of "com.distriqt.facebook.Core.ane" and "com.distriqt.facebook.Login.ane" and disabled the facebook login code in our client.
=> Now I was able to create the build without any errors. :)

4) Next I tried to start our app which worked but unfortunately we still have io errors as soon as we try to load files from the asset pack. I tried to to use an empty root path as well as the name of the asset pack.
```
Stream Error. URL: app:/assetsExported/android-etc/UI/shared/assets/LibraryAtlas.xml
Stream Error. URL: app:/MainAssetPack/assetsExported/android-etc/UI/shared/assets/LibraryAtlas.xml
```

This would be the path of asset pack in the AndroidStudioProject:
```
AndroidStudioProject\MainAssetPack\src\main\assets
```
Does the PlayAssetDelivery.ane provide some special classes that need to be used to access files from the asset-pack
or is there some kind of internal API hooking going on, so it's still possible to use the URLLoader class?

**Reply by marchbold** [reply]
@ajwfrost Looks as though you have included the appcompat lib (jar) in this ANE but have missed the associated resources?

Can you please use the common dependencies when developing ANEs so we don't have conflicts like this? 

https://github.com/distriqt/ANE-AndroidSupport/tree/master/lib

Let me know if you need to update the version in the lib.

**Reply by ajwfrost** [reply]
@marchbold  I've asked a colleague to look into that part
@dbartsch-bf  ... and we'll also get together a quick guide to using the ANE! Looking at the test app here though, it seems to need you to use the URL that is returned from a call to `PlayAssetDeliveryAne.GetAssetAbsolutePath()` where you pass in the asset pack name followed by the path of the file within the asset pack.
So maybe try:
```
var urlToRequest : URLRequest = new URLRequest( _anePAD.GetAssetAbsolutePath("MainAssetPack", "assetsExported/android-etc/UI/shared/assets/LibraryAtlas.xml") );
```

**Reply by marchbold** [reply]
@ajwfrost Please reach out if you need guidance, ANEs like this with packaged common libs are what we have pushed hard not to have developers build!

**Reply by dbartsch-bf** [reply]
@ajwfrost Thanks for the example code! I will try that and let you know if this works.
Also thanks to marchbold for the support regarding the common libs!

**Reply by dbartsch-bf** [reply]
@ajwfrost I'm currently getting a null reference error each time I'm calling the GetAssetAbsolutePath-method of the PlayAssetDeliveryAne instance.
I also tried to call initAssetDelivery first but that call also fails with a null reference error..

```
TypeError: Error #1009: Cannot access a property or method of a null object reference.
at com.harman.air.extension::PlayAssetDeliveryAne/GetAssetAbsolutePath()
at mobile::PlayAssetDeliveryCtrl/getMainAssetPackURLRequest()
```

Here is the code:
```
private var mPlayAssetDeliveryAne:PlayAssetDeliveryAne;
// ...
public function init():void
{
	mPlayAssetDeliveryAne = new PlayAssetDeliveryAne();
	mPlayAssetDeliveryAne.initAssetDelivery();
}

public function getMainAssetPackURLRequest(path:String):URLRequest
{
	var apPath:String = mPlayAssetDeliveryAne.GetAssetAbsolutePath("MainAssetPack", path);
	return new URLRequest(apPath);
}
```

The code is executed in our handler of the Event.ADDED_TO_STAGE by the way.
Is there something else that I'm missing? Maybe in the application descriptor file?

**Reply by ajwfrost** [reply]
@dbartsch-bf sorry, we'll need to come back to you with some usage instructions for this...!
Is it an install-time one? maybe use `_ane.LoadInstallTimeAssetANE()`?
We'll get some better documentation soon!!

**Reply by dbartsch-bf** [reply]
@ajwfrost Additional usage instructions would be great as I can only see the class definition of the PlayAssetDeliveryAne class in Flash-Develop at the moment. 👍 

Yes, we use a single "install-time"-asset-pack.
```
<assetPacks>
	<assetPack id="MainAssetPack" delivery="install-time" folder="MainAssetPack" />
</assetPacks>	
```

I just tried to use the LoadInstallTimeAssetANE-method but this leads to the same null reference error:
Code:
```
mPlayAssetDeliveryAne.LoadInstallTimeAssetANE("MainAssetPack");
```
Error:
```
TypeError: Error #1009: Cannot access a property or method of a null object reference.
at com.harman.air.extension::PlayAssetDeliveryAne/LoadInstallTimeAssetANE()
at mobile::PlayAssetDeliveryCtrl/init()
```
Is there maybe a problem with the initialization of the extension context?

**Reply by ajwfrost** [reply]
Yes it looks like there's a problem with the initialization: can you please confirm what platform/device you're using to test on? and how are you building/deploying -> is this via an App Bundle and then deploying it locally via bundletool/ADT?

Wondering if we can get your XML file and command-line arguments to try to reproduce this here... below is a link if you're able to e.g. zip up your project and we can try to reproduce it.
https://transfer.harman.com/requests/KzG0tnyZuarRSQp5oYEDmV

thanks

**Reply by dbartsch-bf** [reply]
@ajwfrost Thanks for the help!
I'm currently testing with an Android-x86-Emulator with API 29. Yes, I enabled the
app bundle creation in the adt.cfg and we deploy the aab via the bundletool.
I'm on vacation for the rest of the day but I should be able to provide some snippets
of our application descriptor and build scripts on Monday. Have a nice weekend!

**Reply by dbartsch-bf** [reply]
Hi @ajwfrost!
I was able to upload a simple example air project by the transfer link you provided yesterday
where you should be able to reproduce the ane error.
Just let me know if you have any questions regarding the project.

**Reply by ajwfrost** [reply]
@dbartsch-bf  sorry for the delay here, we did find an issue with the x86 component of the ANE.. so the below version should now work for this...
https://transfer.harman.com/message/FqrQLqZCJiz1URcSvZ61c7
Were you able to get it working properly on real (arm-based) devices?
thanks

**Comment by ajwfrost** [comment]
Some additional information here:

The additional entries need to be added into the application descriptor manifestAdditions:
```
<application android:enabled="true">
    <activity
            android:name="com.google.android.play.core.common.PlayCoreDialogWrapperActivity"
            android:enabled="false"
            android:exported="false"
            android:stateNotNeeded="true" />
    <service
            android:name="com.google.android.play.core.assetpacks.AssetPackExtractionService"
            android:enabled="false"
            android:exported="true" />
    <service
            android:name="com.google.android.play.core.assetpacks.ExtractionForegroundService"
            android:enabled="false"
            android:exported="false" />
</application>
```

And then to use the ANE:

```
_ane = new PlayAssetDeliveryAne();
status =_ane.initAssetDelivery() //this  should be called first and should return status success

// For  loading install time asset 
status =_ane.LoadInstallTimeAssetANE("asset-name");//this should return status success, here asset-name is the name of the asset file  inside install time asset Pack

// to download or install Fast-follow and Demand Delivery  assetpack  
status =_ane.FetchAsset("assetPackName"); // This method returns  the AssetPackStates initial download states and sizes.

// To check the status of assetPack:
status = _ane.GetAssetStatus("assetPackName");//this method return the status of assetpack  6- Cancelled,4-Completed,2-Downloading,5-failed,8-Not_installed,1-pending,3-Transferring,0-unknown

// to get absolute path for an asset:
status =_ane.GetAssetAbsolutePath("assetPackName", “path");//this method returns the absolute path only  if the assetpack state is completed -4 ,otherwise it returns null.path refers to path of the file with asset pack
```

Hope that helps: it may be those manifest entries that were missing, or perhaps call that "GetAssetStatus" to confirm the status of the asset pack prior to accessing any data...

thanks

**Reply by dbartsch-bf** [reply]
@ajwfrost Thanks for the infos! 👍 
I added the activity and the two services to the application descriptor and called initAssetDelivery first but unfortunately I still get the same error. Also checked the AndroidManifest.xml in the "base/manifest" folder of our AAB and the new entries were there.
````
TypeError: Error #1009: Cannot access a property or method of a null object reference.
        at com.harman.air.extension::PlayAssetDeliveryAne/initAssetDelivery()
        at mobile::PlayAssetDeliveryCtrl/init()
        at com.traviangames.railnation::MainGoogle/onAddedToStage()
````
Maybe it would help to add some additional traces to see were the null reference access occurs.

**Comment by sjabberwocky** [comment]
Hi @ajwfrost!
Is there an expectation of when the extension will be ready for release?

**Reply by ajwfrost** [reply]
@sjabberwocky @dbartsch-bf Sorry, I had had an action to upload the latest version of this but had missed it :-(  

Below is a link to the ANE which is updated a bit, basically with a different method to access a file that was done at install-time vs the assets downloaded at run-time. We're looking to see whether we could expose the same API for the run-time/delayed-loaded asset pack content, but actually you can just open it using normal File operations so that should be okay I hope..

https://transfer.harman.com/message/Btl0ZVOr2XSARVq7A2Zxul

You still need the application descriptor additions, and the com.google.android.play ANE, per the above. Please let us know of any issues or questions, we'll aim to get this properly documented and submitted to the APM shortly!

thanks

**Reply by marchbold** [reply]
The API looks much better now @ajwfrost ! 👍 

However would suggest a custom event rather than using the `StatusEvent`, `PayAssetDeliveryEvent` with `status` and `assetPackName` properties?

**Reply by marchbold** [reply]
I've setup a quick docs page here to start: https://github.com/airsdk/ANE-PlayAssetDelivery/wiki

**Comment by AdamSun** [comment]
Hi @ajwfrost ，we are testing your new PlayAssetDelivery ANE from https://transfer.harman.com/message/Btl0ZVOr2XSARVq7A2Zxul  It seems that you have changed some code, for example the "OpenInstallTimeAsset" Funciton to replace "LoadInstallTimeAsset".We found that the swc library is new  ,but the harmanPlayAssetDelivery.jar is still the old one.Can you delivery a new ANE to fix this problem?Thanks!

**Reply by ajwfrost** [reply]
Not sure what happened there..! can you please try the one I'll just attach here....
thanks

[PlayAssetDelivery.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7553578/PlayAssetDelivery.zip)

**Reply by AdamSun** [reply]
It works!Thanks!

**Reply by ajwfrost** [reply]
Great! ... we'll be making a slight update following the earlier suggestion from @marchbold and will then get this published..
thanks

**Reply by AdamSun** [reply]
Great job！We suggest that the usage description of the ANE and the download link can be updated in the AIR SDK release note file so that others can find it easily...Thanks for your great work again^-^
